### PR TITLE
[libc] lift app's definition out of startup

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
AppProperties are not target specific hence this patch lifts app from a header library to an object library. In overlay mode, this also provides a way to check if libc is loaded by its own CRT.

### CODE REVIEW

**Comment by SchrodingerZhu** [comment]
@michaelrj-google PTAL

**Comment by SchrodingerZhu** [comment]
Will be rebased.

File: libc/config/linux/app.h [hunk 1]
```diff
@@ -18,21 +18,23 @@ namespace LIBC_NAMESPACE {
...
   // PT_TLS header.
-  uintptr_t init_size;
+  uintptr_t init_size = 0;
 
   // The alignment of the TLS layout. It assumed that the alignment
   // value is a power of 2.
-  uintptr_t align;
+  uintptr_t align = 0;
+
+  constexpr TLSImage() = default;
```
Comment (michaelrj-google) [inline libc/config/linux/app.h:37]: is this explicit constructor necessary? If it is, it should be marked `LIBC_INLINE`.

File: libc/config/linux/app.cpp [hunk 2]
```diff
@@ -0,0 +1,13 @@
+//===-- Implementation file for linux applications classes  -----*- C++ -*-===//
```
Comment (michaelrj-google) [inline libc/config/linux/app.cpp:1]: why was this file added?

**Review by michaelrj-google — COMMENTED** [review]
This patch seems fine, I'm just not sure why these structs need default values for all their members, given that they are immediately initialized in the real code. This seems like it would have either no effect, or it would have a slight performance penalty.

File: libc/config/linux/app.cpp (hunk 2 above)
Comment (SchrodingerZhu) [reply libc/config/linux/app.cpp:1]: `AppProperties app` was defined in `start.cpp`; I moved its definition here.

File: libc/config/linux/app.h (hunk 1 above)
Comment (SchrodingerZhu) [reply libc/config/linux/app.h]: I am just to unify the style for the default constructor. It seems that this file mark an explicit default constructor for `TLSDescriptor` before.

**Comment by SchrodingerZhu** [comment]
@michaelrj-google Default values are needed in overlay mode, where the `libc` is not loaded by its loader such that `AppProperties` is not available. For example, I may need to check if `auxv_ptr` is `nullptr` for `getauxval` and fallback to syscall/fs based approaches.

**Comment by SchrodingerZhu** [comment]
a gentle ping~

**Comment by SchrodingerZhu** [comment]
this needs to be refactored.

**Comment by SchrodingerZhu** [comment]
This is no longer needed.
